README.md

# SHMT (Static Hash Map Table)

SHMT is an implementation of a very fast key-value read-only
hash map table for PHP7.

We have developed SHMT as a faster, dependency-free replacement for the PECL
[CHDB](https://pecl.php.net/package/chdb) extension.

## Features

SHMT

* is written in C.
* creates and uses its own memory-mapped binary file. This
  enable it to cache and share the loaded pages of the
  file across multiple processes.
* has an extremely fast initial load, regardless of the
  size of the binary file.
* internally implements a "perfect hash function" and
  guarantees O(1) lookup time in the worst cases.
* internally uses the very fast "[MurmurHash3](https://en.wikipedia.org/wiki/MurmurHash)"
  hashing algorithm.
* doesn't require any external libraries.
* is PHP 7 ready.


## Limitations

* Supported maximum number of data array elements is
  2^26 (67,108,864 on 32 bit systems) and 2^31 (2,147,483,648
  on 64 bit systems).
* The data array keys and values are always cast to string.
* Data files cannot be exchanged between 32 bit and 64 bit 
  systems or systems with different endianness.
* The code compiles and runs on Linux systems. Other platforms
  have not been tested.


## PHP Class

```
public static boolean SHMT::create(string $filename, array $array)
```

* Creates a SHMT from the `$array` and writes it into the file `$filename`
* Returns `true` on success
* Throws exceptions on errors


```
public SHMT::__construct(string $filename)
```

* Constructs a new SHMT object and maps the SHMT file `$filename` into memory
* Throws exceptions on errors


```
public (string|null) SHMT::get(string $string)
```

* Attempts to find the value stored under the key `$string`
* Returns the value if the key `$string` exists, otherwise `null`


### Example

Create a SHMT:

```
$array = [
	'str_key' => 'test',
	123456789 => 12345
	-1 => ''
];

SHMT::create($filename = 'map.shmt', $array);
```

Read from a SHMT:

```
$shmt = new SHMT($filename);

echo $shmt->get('str_key');    // string(4) "test"
echo $shmt->get(123456789);    // string(5) "12345"
echo $shmt->get('123456789');  // string(5) "12345"
echo $shmt->get(-1);           // string(0) ""
echo $shmt->get('abc_xyz');    // NULL
```


Changelog

1.0.1dev

 * Fixed: Memory leaks in the `SHMT::create()` method (`7ea1490`)
 * Fixed: The order of the `include`s (`cc9afee`)

 * Improved: Using of appropriate return macros makes superfluous `zval`s initialization unnecessary (`a71018e`)

1.0.0

 * Improved: Trivial documentation and comment fixes (`59e63a3`)
 * Improved: More links and a note about the platform in the `README.md` (`fb1aa69`)


shmt.c
#include <php.h>
#include <Zend/zend_exceptions.h>

#include <fcntl.h>
#include <sys/mman.h>
#include <sys/stat.h>

#include "shmt.h"

/* ==================================================================================================== */

struct _shmtHead shmtCreateHeader(uint32_t size)
{
	struct _shmtHead shmtHead;

	memset(&shmtHead, 0, sizeof(struct _shmtHead));

	shmtHead.mark		= SHMT_MARK;
	shmtHead.version	= SHMT_FILE_VERSION;
	shmtHead.system		= SIZEOF_SIZE_T;
	shmtHead.mask		= _shmtGetMapSize(size) - 1;
	shmtHead.fileSize	= 0;
	shmtHead.hashSize	= (shmtHead.mask + 1) * sizeof(struct _shmtHash);
	shmtHead.mapSize	= shmtHead.hashSize + ((shmtHead.mask + 1) * sizeof(struct _shmtItem));

	return shmtHead;
}

int shmtCreateMapTable(struct _shmtHash **pMap, struct _shmtItem **pTbl, struct _shmtHead shmtHead)
{
	if ((*pMap = (struct _shmtHash *)emalloc(shmtHead.mapSize)) == NULL) {
		return 0;
	}

	/* Assign the forwarded address */
	*pTbl = (struct _shmtItem *)((char *)*pMap + shmtHead.hashSize);

	/* Fill the pMap and pTbl with the empty-index flags */
	memset(((struct _shmtHash *)(*pMap)), UINT32_MAX, shmtHead.mapSize);

	return 1;
}

int shmtWriteItem(struct _shmtCreatorItem *src, struct _shmtItem *shmtItem, FILE *file)
{
	if ((shmtItem->key_pos = ftell(file)) <= 0) {
		return 0;
	}

	shmtItem->key_len = src->key_len;
	shmtItem->val_len = src->val_len;

	if (src->key_len) {
		/* Check for site_t overflow */
		if ((shmtItem->key_pos + shmtItem->key_len) < shmtItem->key_pos) {
			return 0;
		} else if (fwrite(src->key, 1, src->key_len, file) != (size_t)src->key_len) {
			return 0;
		}
	}

	if (src->val_len) {
		/* Check for site_t overflow */
		if ((shmtItem->key_pos + shmtItem->key_len + shmtItem->val_len) <= shmtItem->key_pos) {
			return 0;
		} else if (fwrite(src->val, 1, src->val_len, file) != (size_t)src->val_len) {
			return 0;
		}
	}

	return 1;
}

int shmtReadTable(const char *path, struct _shmtHead **shmt)
{
	struct stat	sizeCheck;
	int			file;

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	if ((file = open(path, O_RDONLY)) < 0) {
		return shmtCleanupReader(-1, "SHMT: Cannot open the file");
	} else if (fstat(file, &sizeCheck) != 0) {
		return shmtCleanupReader(file, "SHMT: Cannot read the file size");
	} else if ((size_t)sizeCheck.st_size > SIZE_MAX) {
		return shmtCleanupReader(file, "SHMT: Given file is out of the supported file size range");
	} else if ((size_t)sizeCheck.st_size < sizeof(struct _shmtHead)) {
		return shmtCleanupReader(file, "SHMT: Cannot read the header");
	} else if ((*shmt = (struct _shmtHead *)mmap(NULL, (size_t)sizeCheck.st_size, PROT_READ, MAP_SHARED, file, 0)) == MAP_FAILED) {
		return shmtCleanupReader(file, "SHMT: Cannot map the file to the memory");
	} else if ((size_t)sizeCheck.st_size != (*shmt)->fileSize) {
		return shmtCleanupReader(file, "SHMT: Corrupt SHMT file detected");
	} else if ((*shmt)->system != SIZEOF_SIZE_T) {
		return shmtCleanupReader(file, "SHMT: Incompatible file given");
	} else if (((*shmt)->mark != SHMT_MARK) || ((*shmt)->version != SHMT_FILE_VERSION)) {
		return shmtCleanupReader(file, "SHMT: Unsupported file given");
	}

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	return shmtCleanupReader(file, NULL);
}

int shmtCleanupReader(int file, const char *message)
{
	if (file >= 0) {
		close(file);
	}

	if (message != NULL) {
		zend_throw_exception(zend_ce_exception, message, 0);
		return 0;
	}

	return 1;
}

void shmtCleanup(struct _shmtHead *head, struct _shmtHash **map, FILE **file, const char *path, struct _shmtCreatorItem **it, struct _shmtCreatorList **li, const char *message)
{
	struct _shmtCreatorList	*pLItem;
	struct _shmtCreatorItem	*pCItem;
	size_t					idx, iterator;

	if (map != NULL) {
		efree(*map);
	}

	if (file != NULL) {
		fclose(*file);
	}

	if (path != NULL) {
		remove(path);
	}

	if (it != NULL) {
		if ((li != NULL) && (head != NULL)) {
			for (iterator = 0; iterator <= head->mask; iterator++) {
				pLItem = ((struct _shmtCreatorList *)(*li)) + iterator;

				if (pLItem->num <= 0) {
					break; /* No more items available, nothing to free any more */
				}

				idx = pLItem->idx;

				if (pLItem->num > 1) {
					while (idx != UINT32_MAX) {
						pCItem = ((struct _shmtCreatorItem *)(*it)) + idx;

						efree(pCItem->key);
						efree(pCItem->val);

						idx = pCItem->next;
					}
				} else {
					pCItem = ((struct _shmtCreatorItem *)(*it)) + idx;

					efree(pCItem->key);
					efree(pCItem->val);
				}
			}
		}

		efree(*it);
	}

	if (li != NULL) {
		efree(*li);
	}

	if (message != NULL) {
		zend_throw_exception(zend_ce_exception, message, 0);
	}
}

void shmtFree(struct _shmtHead **shmt)
{
	if (*shmt && (*shmt != MAP_FAILED)) {
		munmap(*shmt, (*shmt)->fileSize);
	}
}

inline uint32_t _shmtGetMapSize(uint32_t size)
{
	if (size < SHMT_MIN_T_SIZE) {
		size = SHMT_MIN_T_SIZE;
	} else if (size > SHMT_MAX_T_SIZE) {
		return 0;
	}

	size -= 1;
	size |= (size >>  1);
	size |= (size >>  2);
	size |= (size >>  4);
	size |= (size >>  8);
	size |= (size >> 16);

	return size + 1;
}

/* ==================================================================================================== */

/**
 * MurmurHash3 implementation
 * --------------------------
 * 
 * MurmurHash3 was written by Austin Appleby, and is placed in the public
 * domain. The author hereby disclaims copyright to this source code.
 */

#define ROT32(x, y) (x << y) | (x >> (32 - y))

uint32_t shmtGetHash(const void *key, size_t len, uint32_t seed)
{
	const uint8_t *data = (const uint8_t *)key;
	const int nblocks = len / 4;
	int i;

	uint32_t hash = seed, c1 = 0xcc9e2d51, c2 = 0x1b873593, k1;
	const uint32_t *blocks = (const uint32_t *)(data + nblocks * 4);

	for(i = -nblocks; i; i++) {
		k1 = blocks[i];

		k1 *= c1;
		k1 = ROT32(k1, 15);
		k1 *= c2;

		hash ^= k1;
		hash = ROT32(hash, 13); 
		hash = hash * 5 + 0xe6546b64;
	}

	const uint8_t *tail = (const uint8_t *)(data + nblocks * 4);
	k1 = 0;

	switch(len & 3)
	{
		case 3: k1 ^= tail[2] << 16;
		case 2: k1 ^= tail[1] << 8;
		case 1: k1 ^= tail[0];
				k1 *= c1; k1 = ROT32(k1, 15); k1 *= c2; hash ^= k1;
	};

	hash ^= len;

	hash ^= hash >> 16;
	hash *= 0x85ebca6b;
	hash ^= hash >> 13;
	hash *= 0xc2b2ae35;
	hash ^= hash >> 16;

	return hash;
}


php_shmt.c
#include <php.h>
#include <Zend/zend_exceptions.h>
#include <ext/standard/info.h>

#include "php_shmt.h"
#include "shmt.h"

/* ==================================================================================================== */

zend_class_entry *ceSHMT;
zend_object_handlers SHMT_handlers;

inline SHMT_object *SHMT_fetch(zend_object *object)
{
	return (SHMT_object *)((char *)(object) - XtOffsetOf(SHMT_object, std));
}

zend_object *SHMT_new(zend_class_entry *ce)
{
	SHMT_object *intern = (SHMT_object *)ecalloc(1, sizeof(SHMT_object) + zend_object_properties_size(ce));

	intern->shmt	= NULL;
	intern->map		= NULL;
	intern->tbl		= NULL;

	zend_object_std_init(&intern->std, ce);
	object_properties_init(&intern->std, ce);
	intern->std.handlers = &SHMT_handlers;

	return &intern->std;
}

void SHMT_free(zend_object *object)
{
	SHMT_object *intern = SHMT_fetch(object);

	zend_object_std_dtor(&intern->std);

	shmtFree(&intern->shmt);
}

/* ==================================================================================================== */

ZEND_BEGIN_ARG_INFO_EX(arginfo_SHMT__construct, 0, 0, 1)
	ZEND_ARG_TYPE_INFO(0, filename, IS_STRING, 0)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_SHMT_get, 0, 0, 1)
	ZEND_ARG_TYPE_INFO(0, string, IS_STRING, 0)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_SHMT_create, 0, 0, 2)
	ZEND_ARG_TYPE_INFO(0, filename, IS_STRING, 0)
	ZEND_ARG_TYPE_INFO(0, array, IS_ARRAY, 0)
ZEND_END_ARG_INFO()

/* ==================================================================================================== */

PHP_METHOD(SHMT, __construct)
{
	SHMT_object	*object = SHMT_fetch(Z_OBJ_P(getThis()));
	char		*path;
	size_t		pathLen;

	if (SUCCESS != zend_parse_parameters(ZEND_NUM_ARGS(), "s", &path, &pathLen)) {
		return;
	} else if (!shmtReadTable(path, &object->shmt)) {
		return;
	}

	/* Assign the pointer to the right addresses/offsets */
	object->map = (struct _shmtHash *)((char *)object->shmt + sizeof(struct _shmtHead));
	object->tbl = (struct _shmtItem *)((char *)object->map + object->shmt->hashSize);
}

PHP_METHOD(SHMT, get)
{
	SHMT_object			*object;
	char				*key;
	size_t				keyLen;
	uint32_t			hash;
	struct _shmtItem	*shmtItem;
	struct _shmtHash	*shmtHash;

	if (SUCCESS != zend_parse_parameters(ZEND_NUM_ARGS(), "s", &key, &keyLen)) {
		return;
	}

	object = SHMT_fetch(Z_OBJ_P(getThis()));

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	hash = shmtGetHash(key, keyLen, 0);
	shmtHash = &object->map[hash & object->shmt->mask];

	if (shmtHash->seed == UINT32_MAX) {
		/* Direct hit */
		shmtItem = &shmtHash->hit;
	} else {
		/* Resolved item */
		hash = shmtGetHash(key, keyLen, shmtHash->seed);
		shmtItem = &object->tbl[hash & object->shmt->mask];
	}

	if ((shmtItem->key_pos != SIZE_MAX) && (keyLen == shmtItem->key_len)) {
		if ((memcmp((const void *)((char *)object->shmt + shmtItem->key_pos), key, keyLen) == 0)) {
			RETURN_STRINGL((char *)object->shmt + shmtItem->key_pos + shmtItem->key_len, shmtItem->val_len);
		}
	}

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	/* No match found -> return a NULL value */
	RETURN_NULL();
}

PHP_METHOD(SHMT, create)
{
	char				*path;
	size_t				pathLen;
	zval				*data, currKey, *currVal;
	HashTable			*htData;
	HashPosition		hpPos;
	uint32_t			iCount, iItemIndex, iterator = 0;
	FILE				*pFile;
	struct _shmtHead	shmtHead;
	struct _shmtHash	*pMap;
	struct _shmtItem	*pTbl, *shmtItem;

	if (SUCCESS != zend_parse_parameters(ZEND_NUM_ARGS(), "sa", &path, &pathLen, &data)) {
		return;
	}

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	htData = Z_ARRVAL_P(data);
	zend_hash_internal_pointer_reset_ex(htData, &hpPos);

	iCount = zend_hash_num_elements(htData);

	if (iCount < 1) {
		return shmtCleanup(NULL, NULL, NULL, NULL, NULL, NULL, "SHMT: Empty data array given");
	} else if ((pFile = fopen(path, "w")) == NULL) {
		return shmtCleanup(NULL, NULL, NULL, NULL, NULL, NULL, "SHMT: Cannot open the file");
	}

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	shmtHead = shmtCreateHeader(iCount);

	if (shmtHead.mask == UINT32_MAX) {
		return shmtCleanup(&shmtHead, NULL, &pFile, path, NULL, NULL, "SHMT: Number of the given elements is out of range");
	}

	if ((fwrite(&shmtHead, 1, sizeof(shmtHead), pFile)) < sizeof(shmtHead)) {
		return shmtCleanup(&shmtHead, NULL, &pFile, path, NULL, NULL, "SHMT: Cannot write the header");
	}

	if (!shmtCreateMapTable(&pMap, &pTbl, shmtHead)) {
		return shmtCleanup(&shmtHead, NULL, &pFile, path, NULL, NULL, "SHMT: Cannot create the table");
	}

	if (fseek(pFile, shmtHead.mapSize, SEEK_CUR)) {
		return shmtCleanup(&shmtHead, &pMap, &pFile, path, NULL, NULL, "SHMT: Unexpected internal \"seek\" error");
	}

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	uint32_t n, newHashVal, newHashMod, newSeed;
	struct _shmtCreatorItem	*pCItem = NULL, *pCItems = NULL;
	struct _shmtCreatorList	*pLItem = NULL, *pLItems = NULL;

	if ((pCItems = (struct _shmtCreatorItem *)emalloc((shmtHead.mask + 1) * sizeof(struct _shmtCreatorItem))) == NULL) {
		return shmtCleanup(&shmtHead, &pMap, &pFile, path, NULL, NULL, "SHMT: Unexpected internal \"malloc\" error");
	}

	if ((pLItems = (struct _shmtCreatorList *)emalloc((shmtHead.mask + 1) * sizeof(struct _shmtCreatorList))) == NULL) {
		return shmtCleanup(&shmtHead, &pMap, &pFile, path, &pCItems, NULL, "SHMT: Unexpected internal \"malloc\" error");
	}

	memset(pCItems, 0, (shmtHead.mask + 1) * sizeof(struct _shmtCreatorItem));
	memset(pLItems, 0, (shmtHead.mask + 1) * sizeof(struct _shmtCreatorList));

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	while ((currVal = zend_hash_get_current_data_ex(htData, &hpPos)) != NULL) {
		zend_hash_get_current_key_zval_ex(htData, &currKey, &hpPos);

		convert_to_string(currVal);
		convert_to_string(&currKey);

		iItemIndex		= iterator++;
		pCItem			= ((struct _shmtCreatorItem *)(pCItems)) + iItemIndex;

		pCItem->hash	= shmtGetHash(Z_STRVAL(currKey), Z_STRLEN(currKey), 0);
		pCItem->hash	= (uint32_t)(pCItem->hash & shmtHead.mask);

		pCItem->key		= estrndup(Z_STRVAL(currKey), Z_STRLEN(currKey));
		pCItem->key_len	= Z_STRLEN(currKey);
		pCItem->val		= estrndup(Z_STRVAL_P(currVal), Z_STRLEN_P(currVal));
		pCItem->val_len	= Z_STRLEN_P(currVal);

		pLItem			= ((struct _shmtCreatorList *)(pLItems)) + pCItem->hash;

		if (!pLItem->num) {
			pLItem->idx = UINT32_MAX;
		}

		pCItem->next	= pLItem->idx;
		pLItem->idx		= iItemIndex;
		pLItem->num++;

		zval_dtor(&currKey);
		zend_hash_move_forward_ex(htData, &hpPos);
	}

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	qsort((void *)pLItems, shmtHead.mask + 1, sizeof(struct _shmtCreatorList), shmtSortCmp);

	for (iterator = 0; iterator <= shmtHead.mask; iterator++) {
		pLItem = ((struct _shmtCreatorList *)(pLItems)) + iterator;

		/* Ready, all items are written. */
		if (pLItem->num <= 0) {
			break;
		}

		iItemIndex = pLItem->idx;

		/* Resolve collisions */
		if (pLItem->num > 1) {
			iCount = 0;
			newSeed = 1;

			uint32_t aTempMod[pLItem->num];
			memset(&aTempMod, UINT32_MAX, sizeof(aTempMod));

			while (iItemIndex != UINT32_MAX) {
				pCItem = ((struct _shmtCreatorItem *)(pCItems)) + iItemIndex;

				newHashVal = shmtGetHash(pCItem->key, pCItem->key_len, newSeed);
				newHashMod = newHashVal & shmtHead.mask;

				shmtItem = ((struct _shmtItem *)(pTbl)) + newHashMod;

				if (shmtItem->key_pos != SIZE_MAX) {
					newSeed++;
					memset(&aTempMod, UINT32_MAX, sizeof(aTempMod));
					iItemIndex = pLItem->idx;
					iCount = 0;
				} else {
					for (n = 0; n <= iCount; n++) {
						if (aTempMod[n] == UINT32_MAX) {
							aTempMod[iCount] = newHashMod;
							iItemIndex = pCItem->next;
							iCount++;
							break;
						}

						if (aTempMod[n] == newHashMod) {
							newSeed++;
							memset(&aTempMod, UINT32_MAX, sizeof(aTempMod));
							iItemIndex = pLItem->idx;
							iCount = 0;
							break;
						}
					}
				}
			}

			pMap[pCItem->hash].seed = newSeed;

			iItemIndex = pLItem->idx;
			for (n = 0; n < pLItem->num; n++) {
				pCItem = ((struct _shmtCreatorItem *)(pCItems)) + iItemIndex;
				shmtItem = ((struct _shmtItem *)(pTbl)) + aTempMod[n];

				if (!shmtWriteItem(pCItem, shmtItem, pFile)) {
					return shmtCleanup(&shmtHead, &pMap, &pFile, path, &pCItems, &pLItems, "SHMT: Unexpected internal \"write\" error");
				}

				iItemIndex = pCItem->next;
			}
		} else {
			/* Collision free items are written directly into the pMap (not pTbl) array. */
			pCItem = ((struct _shmtCreatorItem *)(pCItems)) + iItemIndex;

			if (!shmtWriteItem(pCItem, &pMap[pCItem->hash].hit, pFile)) {
				return shmtCleanup(&shmtHead, &pMap, &pFile, path, &pCItems, &pLItems, "SHMT: Unexpected internal \"write\" error");
			}
		}
	}

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	if (fseek(pFile, sizeof(shmtHead), SEEK_SET) || ((fwrite(pMap, 1, shmtHead.mapSize, pFile)) < shmtHead.mapSize)) {
		return shmtCleanup(&shmtHead, &pMap, &pFile, path, &pCItems, &pLItems, "SHMT: Cannot write the table");
	}

	if (
		(fseek(pFile, 0, SEEK_END) != 0) ||
		((shmtHead.fileSize = ftell(pFile)) <= 0) ||
		(fseek(pFile, (char *)&shmtHead.fileSize - (char *)&shmtHead, SEEK_SET) != 0) ||
		((fwrite(&shmtHead.fileSize, 1, sizeof(shmtHead.fileSize), pFile)) < sizeof(shmtHead.fileSize))
	) {
		return shmtCleanup(&shmtHead, &pMap, &pFile, path, &pCItems, &pLItems, "SHMT: Unexpected internal \"finalize\" error");
	}

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

	shmtCleanup(&shmtHead, &pMap, &pFile, NULL, &pCItems, &pLItems, NULL);

	RETURN_TRUE;
}

/* ==================================================================================================== */

zend_function_entry SHMT_methods[] = {
	PHP_ME(SHMT, __construct, arginfo_SHMT__construct, ZEND_ACC_PUBLIC | ZEND_ACC_CTOR)
	PHP_ME(SHMT, get, arginfo_SHMT_get, ZEND_ACC_PUBLIC)
	PHP_ME(SHMT, create, arginfo_SHMT_create, ZEND_ACC_PUBLIC | ZEND_ACC_STATIC)
	PHP_FE_END
};

/* ==================================================================================================== */

static PHP_MINIT_FUNCTION(shmt)
{
	zend_class_entry ce;

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  */

	memcpy(&SHMT_handlers, zend_get_std_object_handlers(), sizeof(zend_object_handlers));
	SHMT_handlers.offset = XtOffsetOf(SHMT_object, std);
	SHMT_handlers.dtor_obj = zend_objects_destroy_object;
	SHMT_handlers.free_obj = SHMT_free;

	INIT_CLASS_ENTRY(ce, "SHMT", SHMT_methods);
	ce.create_object = SHMT_new;
	ceSHMT = zend_register_internal_class_ex(&ce, NULL);

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  */

	return SUCCESS;
}

static PHP_MINFO_FUNCTION(shmt)
{
	php_info_print_table_start();
		php_info_print_table_header(2, "SHMT", "enabled");
		php_info_print_table_row(2, "Version", PHP_SHMT_EXTVER);
	php_info_print_table_end();
}

/* ==================================================================================================== */

zend_module_entry shmt_module_entry =
{
	STANDARD_MODULE_HEADER,
	PHP_SHMT_EXTNAME,
	NULL,					/* Functions */
	PHP_MINIT(shmt),
	NULL,					/* MSHUTDOWN */
	NULL,					/* RINIT */
	NULL,					/* RSHUTDOWN */
	PHP_MINFO(shmt),
	PHP_SHMT_EXTVER,
	STANDARD_MODULE_PROPERTIES
};

#ifdef COMPILE_DL_SHMT
	ZEND_GET_MODULE(shmt)
#endif


shmt.h
#ifndef SHMT_H
	#define SHMT_H

	#include <arpa/inet.h>

	/* ==================================================================================================== */

	#define SHMT_MARK			htonl('s' << 24 | 'h' << 16 | 'm' <<  8 | 't')
	#define SHMT_FILE_VERSION	((uint16_t)(100)) /* Example: 1023 corresponds to version 10.23 */
	#define SHMT_MIN_T_SIZE		((uint32_t)(2))

	#if SIZEOF_SIZE_T == 4
		#define SHMT_MAX_T_SIZE	((uint32_t)(1 << 26))
	#elif SIZEOF_SIZE_T == 8
		#define SHMT_MAX_T_SIZE	((uint32_t)(1 << 31))
	#else
		#error "Unknown SIZEOF_SIZE_T"
	#endif

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
	/* STRUCTURES */

	struct _shmtCreatorItem
	{
		char		*key;
		uint32_t	key_len;
		char		*val;
		uint32_t	val_len;
		uint32_t	hash;
		uint32_t	next;
	};

	struct _shmtCreatorList
	{
		uint32_t num;				/* The number of items in the current bucket				*/
		uint32_t idx;				/* The index of the last item in the current bucket			*/
	};

	struct _shmtItem
	{
		size_t		key_pos;		/* The offset to the key string								*/
		uint32_t	key_len;		/* The key string length									*/
		uint32_t	val_len;		/* The value string length									*/
	};

	struct _shmtHash
	{
		uint32_t			seed;	/* The alternative seed										*/
		struct _shmtItem	hit;	/* The direct hit											*/
	};

	struct _shmtHead
	{
		uint32_t	mark;			/* The plausibility check mark								*/
		uint16_t	version;		/* The file compatibility check mark						*/
		uint8_t		system;			/* The system compatibility check mark (32/64 bit)			*/
		uint32_t	mask;			/* The hash map mask										*/
		size_t		mapSize;		/* The total hash map table size							*/
		size_t		hashSize;		/* The size of the "hash" part in the hash map				*/
		size_t		fileSize;		/* The file size in bytes (for "mmap()")					*/
	};

	typedef struct _SHMT_object
	{
		struct _shmtHead	*shmt;	/* The "mmap()" pointer to be able to call "munmap()"		*/
		struct _shmtHash	*map;	/* The pointer to the "map" part in the total map table		*/
		struct _shmtItem	*tbl;	/* The pointer to the "tbl" part in the total map table		*/
		zend_object			std;	/* The PHP class object										*/
	} SHMT_object;

	/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
	/* FUNCTIONS */

	struct _shmtHead shmtCreateHeader(uint32_t size);

	int shmtCreateMapTable(struct _shmtHash **pMap, struct _shmtItem **pTbl, struct _shmtHead shmtHead);

	int shmtWriteItem(struct _shmtCreatorItem *src, struct _shmtItem *dest, FILE *file);

	int shmtReadTable(const char *path, struct _shmtHead **shmt);

	int shmtCleanupReader(int file, const char *message);

	void shmtCleanup(struct _shmtHead *head, struct _shmtHash **map, FILE **file, const char *path, struct _shmtCreatorItem **it, struct _shmtCreatorList **li, const char *message);

	void shmtFree(struct _shmtHead **shmt);

	uint32_t _shmtGetMapSize(uint32_t size);

	uint32_t shmtGetHash(const void *key, size_t len, uint32_t seed);

	static inline int shmtSortCmp(const void *a, const void *b)
	{
		if ((*(struct _shmtCreatorList *)a).num > (*(struct _shmtCreatorList *)b).num) {
			return -1;
		} else if ((*(struct _shmtCreatorList *)a).num < (*(struct _shmtCreatorList *)b).num) {
			return  1;
		} else {
			return  0;
		}
	}

#endif /* SHMT_H */
